### ISSUE TITLE
Think Then Translate - Finetuning with generated COT

### ISSUE THREAD

**Issue by TenzinGayche** [OpenPecha/think_and_translate#2]
**We Believe That:**

Incorporating explicit chain-of-thought (CoT) reasoning into the translation process will significantly enhance translation quality compared to direct translation without explicit reasoning. By breaking down the translation task into intermediate steps, we can gain insight into the model's decision-making process and improve its accuracy. Explicit reasoning will also make evaluation more granular and informative, as we can pinpoint areas where the model excels or struggles.

**To Verify That, We Will:**

1.  **Dataset Preparation (Completed):**
    *   We began with a high-quality dataset of 84000.
    *   We utilized Claude to extract or construct approximately 33,000 self-contained sentences or coherent paragraphs from this dataset, ensuring each unit carries complete meaning in isolation.
    *   For each of the 33,000 sentences/paragraphs, we used Claude to generate corresponding chain-of-thought rationales for the translations. Each CoT details the reasoning steps Claude (acting as a translator) takes to arrive at the translated output from the source text.

2.  **Model Training:**
    *   Train a base translation model to generate CoT rationales *before* producing the final translation. This will encourage the model to internalize a reasoning process.
    *   Employ temperature sampling during inference to generate a diverse set of CoT rationales and corresponding translations for each input.
    *   **Translation Quality Ranking:** We will use one of the following methods to rank the generated translations, with the option of combining them being a possibility that can be explored:
        *   **BLEURT:** Use BLEURT scores as a proxy for translation quality. While not a traditional reward model, BLEURT scores will provide an initial ranking of the translations.
        *   **Claude-based Head-to-Head Evaluation with ELO:** Utilize Claude to perform head-to-head comparisons between generated translations, selecting the translation closer to the target translation. Statistical analysis of these comparisons will be used to derive ELO scores for each translation.
    *   Select the best and worst translation pairs (based on either BLEURT or ELO rankings, or potentially a combination of both). These pairs will be used for further training and could potentially form the basis for training a dedicated reward model in the future.

3.  **Difficulty Classification and Training Optimization:**
    *   Employ Claude to classify each training example (source text, CoT, translation) into difficulty levels: hard, medium, and low. This will enable us to analyze model performance across different complexities.
    *   Aim for concise CoT rationales for easier examples and more elaborate reasoning for complex ones, mimicking human translation behavior.

4.  **Iterative Refinement with Direct Preference Optimization (DPO):**
    *   Train the model using DPO with the selected best/worst pairs for a few epochs. Crucially, DPO will be used to optimize *both* the CoT generation and the final translation, ensuring the reasoning process aligns with the desired output.
    *   Generate new translations using the DPO-refined model.
    *   Repeat the selection of best/worst pairs as described in step 2.
    *   Train a new model with these updated pairs. We will perform two iterations of this DPO refinement.

**Test Set Would Be:**

A curated set of 1,000 - 3,000 translation pairs, carefully selected to represent a range of complexities. This size allows for feasible human evaluation in addition to automated metrics. A direct translation model (without CoT) will be evaluated on the same test set for comparative analysis.

**And Measure:**

We will evaluate the CoT translation model against the direct translation baseline using both automated metrics (e.g., BLEU, BLEURT) and human evaluation. The human evaluation will assess fluency, adequacy, and overall translation quality.

**We Are Right If:**

The CoT translation model demonstrates significantly better performance than the direct translation model, as measured by both automated metrics and human evaluations. If successful, this approach can be leveraged to generate high-quality synthetic data for further training and improvement of translation models, contributing to advancements in machine translation.

Inspired by ChatGPT o1, [Thinking LLM](https://arxiv.org/abs/2410.10630)

**Comment by TenzinGayche** [comment]
here is a sample data:

Please translate the following into English: མ་ཕམ་པ་དེ་ནས་མི་དེས་སེམས་ཅན་དེ་དག་ཐམས་ཅད་དེ་བཞིན་གཤེགས་པས་གསུངས་པའི་ཆོས་འདུལ་བ་ལ་ཡང་དག་པར་བཙུད་དེ། ཡང་དག་པར་བཙུད་ནས་འཛིན་དུ་བཅུག་ལ། སེམས་ཅན་དེ་དག་གིས་ཀྱང་ཆོས་དེ་མཉན་ཏེ་ཐོས་ནས་སྐད་ཅིག་ཐང་ཅིག་ཡུད་ཙམ་གཅིག་ལ་འདུས་ཏེ་ཐམས་ཅད་རྒྱུན་དུ་ཞུགས་པར་གྱུར། ལན་ཅིག་ཕྱིར་འོང་བ་དང་ཕྱིར་མི་འོང་བའི་འབྲས་བུ་ཐོབ་པར་གྱུར་པ་ནས། ཟག་པ་ཟད་པ་བསམ་གཏན་པ་བསམ་གཏན་པ་ཆེན་པོ་རྣམ་པར་ཐར་པ་བརྒྱད་ལ་བསམ་གཏན་པའི་དགྲ་བཅོམ་པའི་བར་དུ་གྱུར་ན། Translation:

Let me analyze this complex passage carefully:

Starting with "མ་ཕམ་པ་" - this is "Ajita," a common name/title in Buddhist texts. It's being addressed to someone.

The structure "དེ་ནས་མི་དེས་" indicates a sequential action - "then that man..."

We have "སེམས་ཅན་དེ་དག་ཐམས་ཅད་" referring to "all those beings"

"དེ་བཞིན་གཤེགས་པས་གསུངས་པའི་ཆོས་འདུལ་བ་" - this is referring to the Dharma taught by the Tathāgata. "འདུལ་བ་" specifically relates to guidance or discipline.

The repetition of "ཡང་དག་པར་བཙུད་" emphasizes the introduction or bringing into the teaching.

Temporal markers "སྐད་ཅིག་ཐང་ཅིག་ཡུད་ཙམ་གཅིག་" all indicate very brief time periods - moment, instant, briefly.

Now we're seeing the fruits of practice: "རྒྱུན་དུ་ཞུགས་པ" (stream-enterer), "ལན་ཅིག་ཕྱིར་འོང་" (once-returner), "ཕྱིར་མི་འོང་" (non-returner)

The final section lists advanced attainments: "ཟག་པ་ཟད་པ" (cessation of defilements), "བསམ་གཏན་པ" (meditators), and reference to "རྣམ་པར་ཐར་པ་བརྒྱད" (eight liberations)

Culminates in "དགྲ་བཅོམ་པ" (arhat) status.

This is a classic Buddhist progression text, showing the path from initial introduction to highest attainment, all happening in a remarkably brief time period. Need to maintain technical accuracy while keeping the flow natural.

Final Translation: Then, Ajita, that man introduces those beings to the Dharma's guidance that has been taught by the Tathāgata, and when they have been introduced to the Dharma they listen to it, and having heard it for just one moment, just one instant, briefly, they attain the results of becoming stream entrants, once-returners, and nonreturners, up to becoming arhats, who have brought the defilements to cessation, become meditators, great meditators, and meditate on the eight liberations.

**Comment by ngawangtrinley** [comment]
@TenzinGayche please create a repo for this project and add your jupiter notebook as well as the chain of thought templates generated by Claude. Without this it will be difficult to review

**Comment by 10zinten** [comment]
I think it would be good to include `chrF++` and `TER` metric for the evaluation as well. You can use https://github.com/mjpost/sacrebleu for this.

References:
- chrF++ [paper](https://www.statmt.org/wmt17/pdf/WMT70.pdf)
- TER [paper](https://www.cs.umd.edu/~snover/pub/amta06/ter_amta.pdf)

**Comment by TenzinGayche** [comment]
@10zinten @ngawangtrinley 

### Target Sentence  
**The teacher praised the student because she completed the project ahead of time, showing exceptional dedication and effort.**

### Candidates  
1. **The teacher commended the student for completing the project early, demonstrating remarkable commitment and hard work.**  
2. **The student praised the teacher because he showed exceptional dedication and effort, completing the project ahead of time.**

### Evaluation Metrics  

| Metric  | Sentence One | Sentence Two |
|---------|--------------|--------------|
| BLEURT  | 80           | 69           |
| CHERF   | 39.52        | 72.21        |
| BLEU    | 7.37         | 28.09        |
| TER     | 55.56        | 66.67        |

### Analysis  

By analyzing the translations and their corresponding metric scores, we can draw key insights:  

- **Sentence One** closely matches the target sentence both in meaning and structure. It accurately reflects the relationship between the teacher and the student, as well as the reason for the praise. The high BLEURT score (80) reflects this alignment, as BLEURT evaluates semantic similarity and overall sentence coherence effectively.  

- **Sentence Two**, however, misinterprets the roles, incorrectly stating that the student praised the teacher instead of the reverse. Despite this semantic inaccuracy, Sentence Two achieves a higher BLEU score (28.09 compared to Sentence One’s 7.37). This discrepancy occurs because BLEU focuses on matching individual words and n-grams without understanding the overall meaning of the text.  

The TER (Translation Edit Rate) scores further highlight the issue. While TER measures the amount of editing required to make the candidate sentence match the target, it does not account for semantic mismatches, leading to a relatively close score between the two sentences.

**These results emphasize the limitations of statistical metrics like BLEU and TER in evaluating translation quality. BLEU tends to overvalue surface-level similarities, even in cases where the translation is semantically incorrect (e.g., Sentence Two).**  

**In contrast, neural network-based metrics like BLEURT excel at identifying meaningful differences, as seen in Sentence One’s higher BLEURT score. This makes BLEURT a more reliable choice for assessing translation accuracy, especially for texts that demand a high level of semantic understanding, such as Buddhist texts.**

_Side note: I plan to conduct extensive research on which evaluation metric is most suitable for Buddhist texts, ensuring a balance between technical term accuracy and overall translation quality._

**Comment by TenzinGayche** [comment]
```python
data = [
    {
        "src": "L'enseignant a félicité l'élève parce qu'elle a terminé le projet en avance, montrant une dévotion et un effort exceptionnels.", 
        "mt": "The teacher commended the student for completing the project early, demonstrating remarkable commitment and hard work", 
        "ref": "The teacher praised the student because she completed the project ahead of time, showing exceptional dedication and effort."
    }
]
model_output = model.predict(data, batch_size=8, gpus=1)
# Segment-level scores
print (model_output.scores)

```
Using [xCOMET](https://github.com/Unbabel/COMET) :
| Metric  | Sentence one | Sentence two |
|---------|--------------|--------------|
| xCOMET | 97         | 75         |
| @TenzinGayche  | 98.7         | 10        |

Note: xCOMET-XL is 3B parameter model

**Comment by TenzinGayche** [comment]
Example : 

**Please translate the following into English: བཀའ་སྩལ་པ། བྱམས་པ་ངས་ཆོས་གདགས་པ་རྣམ་པར་གཞག་པ་འདི་ལྟ་སྟེ། མདོའི་སྡེ་དང་། དབྱངས་ཀྱིས་བསྙད་པའི་སྡེ་དང་། ལུང་དུ་བསྟན་པའི་སྡེ་དང་། ཚིགས་སུ་བཅད་པའི་སྡེ་དང་། ཆེད་དུ་བརྗོད་པའི་སྡེ་དང་། གླེང་གཞིའི་སྡེ་དང་། རྟོགས་པ་བརྗོད་པའི་སྡེ་དང་། དེ་ལྟ་བུ་བྱུང་བའི་སྡེ་དང་། སྐྱེས་པའི་རབས་ཀྱི་སྡེ་དང་། ཤིན་ཏུ་རྒྱས་པའི་སྡེ་དང་། རྨད་དུ་བྱུང་བའི་ཆོས་ཀྱི་སྡེ་དང་། གཏན་ལ་བབ་པར་བསྟན་པའི་སྡེ་གང་དག་བྱང་ཆུབ་སེམས་དཔའ་རྣམས་ལ་བཤད་པ་དེ་དག་བྱང་ཆུབ་སེམས་དཔའས་ལེགས་པར་ཐོས། ལེགས་པར་བཟུང་། ཁ་ཏོན་བྱང་བར་བྱས། ཡིད་ཀྱིས་ལེགས་པར་བརྟགས། མཐོང་བས་ཤིན་ཏུ་རྟོགས་པར་བྱས་ནས་དེ་གཅིག་པུ་དབེན་པར་འདུག་སྟེ། Translation:**


Let me work through this carefully...

First, I see "བཀའ་སྩལ་པ།" - this is a formal speech marker, typically translated as "said" or "replied."

Now we have "བྱམས་པ" - that's Maitreya, one of the Buddha's main disciples. This is clearly addressed to him.

"ངས་ཆོས་གདགས་པ་རྣམ་པར་གཞག་པ་" - this is talking about establishing or setting forth the Dharma. The term "གདགས་པ" suggests a systematic presentation or classification.

Hmm, now we have a long list of different types of Buddhist texts or teachings. Let me break this down:
- མདོའི་སྡེ = Sūtra
- དབྱངས་ཀྱིས་བསྙད་པའི་སྡེ = Gāthā
- ལུང་དུ་བསྟན་པའི་སྡེ = Vyākaraṇa
- ཚིགས་སུ་བཅད་པའི་སྡེ = Gāthā (again, but different type)
- ཆེད་དུ་བརྗོད་པའི་སྡེ = Udāna
- གླེང་གཞིའི་སྡེ = Nidāna
- རྟོགས་པ་བརྗོད་པའི་སྡེ = Avadāna
- དེ་ལྟ་བུ་བྱུང་བའི་སྡེ = Itivṛttaka
- སྐྱེས་པའི་རབས་ཀྱི་སྡེ = Jātaka
- ཤིན་ཏུ་རྒྱས་པའི་སྡེ = Vaipulya
- རྨད་དུ་བྱུང་བའི་ཆོས་ཀྱི་སྡེ = Adbhutadharma
- གཏན་ལ་བབ་པར་བསྟན་པའི་སྡེ = Nītārtha

The structure suggests these are being presented as what bodhisattvas should study.

The final section with "གཅིག་པུ་དབེན་པར་འདུག་སྟེ" indicates solitary practice or retreat.

This seems to be about the proper way to study and practice these different types of Buddhist texts.

#Final Translation:
Maitreya, I have set forth the Dharma in the following way: the bodhisattvas should study the Sūtra, Gāthā, Vyākaraṇa, Gāthā, Udāna, Nidāna, Avadāna, Itivṛttaka, Jātaka, Vaipulya, Adbhutadharma, and Nītārtha. The bodhisattvas should study them well, memorize them, recite them, examine them with their minds, and understand them through insight. Then, they should go to a solitary place and practice them.

**Comment by TenzinGayche** [comment]
https://claude.site/artifacts/7b150eb8-7faa-4691-82c1-89125c29f707
